crates/bevy_reflect/src/type_data.rs

```diff
@@ -0,0 +1,166 @@
+use crate::TypeRegistry;
+use core::fmt::{Debug, Formatter};
+use downcast_rs::{impl_downcast, Downcast};
+
+/// Type-erased [`TypeData`].
+///
+/// Type data can be registered to the [`TypeRegistry`] and stored on a type's [`TypeRegistration`].
+///
+/// While type data is often generated using the [`#[reflect_trait]`](crate::reflect_trait) macro,
+/// almost any type that implements [`Clone`] can be considered "type data".
+/// This is because it has a blanket implementation over all `T` where `T: Clone + Send + Sync + 'static`.
+///
+/// See the [crate-level documentation] for more information on type data and type registration.
+///
+/// [`TypeRegistration`]: crate::TypeRegistration
+/// [crate-level documentation]: crate
+pub trait BaseTypeData: Downcast + Send + Sync {
+    fn type_name(&self) -> &'static str;
+    fn clone_type_data(&self) -> Box<dyn BaseTypeData>;
+}
+impl_downcast!(BaseTypeData);
+
+impl<T: 'static + Send + Sync> BaseTypeData for T
+where
+    T: Clone,
+{
+    fn type_name(&self) -> &'static str {
+        std::any::type_name::<T>()
+    }
+
+    fn clone_type_data(&self) -> Box<dyn BaseTypeData> {
+        Box::new(self.clone())
+    }
+}
+
+impl Debug for dyn BaseTypeData {
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        f.write_str(self.type_name())
+    }
+}
+
+/// Trait used for generating data for a type to be stored in a [`TypeRegistry`].
+///
+/// This is used to provide additional information about a type that can be used
+/// dynamically at runtime.
+/// Most often, this data is tied to a trait so that the trait may be used dynamically.
+///
+/// # Naming Convention
+///
+/// The [derive macro] requires that type data be prefixed with `Reflect`.
+/// This is done so that the macro's registrations are less visually noisy.
+/// For example, the type data for [`Default`] is called [`ReflectDefault`],
+/// and would be registered via the derive macro like `#[reflect(Default)]`.
+///
+/// # Example
+///
+/// Let's say we have the following code and want to be able to use the trait dynamically:
+///
+/// ```
+/// # use bevy_reflect::{Reflect};
+/// trait Animal {
+///   fn speak(&self) -> &'static str;
+/// }
```

Comment:
Haha I don't think it's _that_ bad (Rust by Example [uses it](https://doc.rust-lang.org/rust-by-example/trait.html)), but I can understand why we might want to avoid it. I'll replace it with a game-oriented example like you suggested!